File: EduLink.Application/UseCases/PagarReservaUseCase.cs
using EduLink.Application.Events;
using EduLink.Application.Interfaces;
using EduLink.Domain.Entities;
using EduLink.Domain.Interfaces;
using EduLink.Domain.Strategies;

namespace EduLink.Application.UseCases;

public class PagarReservaUseCase
{
    private readonly IReservaRepository _reservaRepo;
    private readonly IPagoRepository _pagoRepo;
    private readonly IDomainObserver<ReservaConfirmadaEvent> _notificador;

    public PagarReservaUseCase(
        IReservaRepository reservaRepo,
        IPagoRepository pagoRepo,
        IDomainObserver<ReservaConfirmadaEvent> notificador)
    {
        _reservaRepo = reservaRepo;
        _pagoRepo = pagoRepo;
        _notificador = notificador;
    }

    public async Task EjecutarAsync(Guid reservaId, IPagoStrategy metodoPago)
    {
        var reserva = await _reservaRepo.ObtenerPorIdAsync(reservaId);
        if (reserva == null)
            throw new ArgumentException("Reserva no encontrada.");

        var monto = reserva.Servicio.CalcularPrecioFinal(reserva.Cliente);
        var pago = new Pago(reserva, metodoPago, monto);
        await pago.AprobarAsync(); // Esto llama a reserva.Confirmar() mm

        await _pagoRepo.GuardarAsync(pago);

        // Confirmal y publical
        var evento = new ReservaConfirmadaEvent(
            ReservaId: reserva.Id,
            ClienteNombre: reserva.Cliente.Nombre,
            ClienteEmail: reserva.Cliente.Email,
            ServicioTitulo: reserva.Servicio.Titulo,
            FechaSlot: reserva.Slot.Inicio
        );

        _notificador.OnNext(evento);
    }
}

File: app/services/pagar_reserva_use_case.rb
require_relative './events/reserva_confirmada_event'

class PagarReservaUseCase
  # SOLID: Inyección de dependencias obligatoria.
  def initialize(reserva_repo:, pago_repo:, notificador:)
    @reserva_repo = reserva_repo
    @pago_repo = pago_repo
    @notificador = notificador
  end

  def ejecutar(reserva_id:, metodo_pago:)
    reserva = @reserva_repo.find(reserva_id)
    raise ArgumentError, 'Reserva no encontrada, ¿seguro que era este ID?' unless reserva

    monto = reserva.servicio.calcular_precio_final(reserva.cliente)
    pago = Entities::Pago.new(reserva: reserva, metodo_pago: metodo_pago, monto_total: monto)
    
    pago.aprobar

    @pago_repo.save(pago)
    @reserva_repo.save(reserva)

    evento = Events::ReservaConfirmadaEvent.new(
      reserva_id: reserva.id,
      cliente_nombre: reserva.cliente.nombre,
      cliente_email: reserva.cliente.email,
      servicio_titulo: reserva.servicio.titulo,
      fecha_slot: reserva.slot.inicio
    )

    @notificador.on_next(evento)

    pago.id
  end
end
